tests/unit/Option/SomeTest.php

```diff
@@ -129,4 +130,41 @@ public function testEquality()
         static::assertFalse($a->equals(Option\some('other')));
         static::assertTrue($a->equals(Option\some('a')));
     }
+
+    public function testZip(): void
+    {
+        $x = Option\some(1);
+        $y = Option\some("hi");
+
+        static::assertTrue(Option\some([1, 'hi'])->equals($x->zip($y)));
+        static::assertTrue(Option\some(['hi', 1])->equals($y->zip($x)));
+    }
+
+    public function testZipWith(): void
+    {
+        $x = Option\some(17.5);
+        $y = Option\some(42.7);
+
+        $point = $x->zipWith($y, static fn($a, $b) => new Fixture\Point($a, $b));
+
+        static::assertTrue(Option\some(new Fixture\Point(17.5, 42.7))->equals($point));
+    }
+
+    /**
+     * @dataProvider provideTestUnzip
+     */
+    public function testUnzip(Option\Option $option, mixed $expectedX, mixed $expectedY): void
+    {
+        [$x, $y] = $option->unzip();
+
+        static::assertSame($expectedX, $x->unwrap());
+        static::assertSame($expectedY, $y->unwrap());
+    }
+
+    private function provideTestUnzip(): iterable
```

Comment:
You'll have to split this one up in 2 different cases: the case where you are able to unzip some([a, b]) and the case where the data is not a tuple of [A, B]

Reply:
I'm not sure what you mean:

The `Option\some(null)->zip(Option\some('hi'))` returns an `Option<array{null, 'hi'}>`, which would be a valid tuple.

`Option\some(1)->zip(Option\some('hi'))` similarly returns an `Option<array{1, 'hi'}>`, another valid tuple.

Also, `Option\some([true, false])` is another way to build the same valid tuple to unzip, `Option<array{true, false}>`.

There are cases where the data is not a valid tuple covered in `NoneTest`. For example, `Option\none()->zip(Option\some(1))` is a valid tuple of `Psl\Option\Option<list{never, 1}>` that will correctly unzip into two `None`s

Reply:
I meant adding tests for the situation where you want to unzip e.g. some(1) - which is not possible by the tuple type assertion that is added in the code.

Reply:
Ow, I was just facing a mutation right now
```diff
$option = $this->option[0];
-Type\shape([Type\mixed(), Type\mixed()])->assert($option);
+
[$a, $b] = $option;
return [some($a), some($b)];
```
and I remembered that I had in a non-pushed version exactly what you mean, sorry!